src/charm.py

```diff
@@ -56,12 +59,24 @@ def __init__(self, *args):
         self.grafana_source_provider = GrafanaSourceProvider(
```

Comment:
Here we need a similar change as in https://github.com/canonical/prometheus-k8s-operator/pull/253

Reply:
So do you mean rename `self.grafana_source_provider` with  `self.grafana_source_consumer`  and moving the `grafana-source` relation from the `provider` side to the `requires` in `metadata.yaml`?

Reply:
Actually, it's the `source_uri` parameter that's needed, maybe...

When `web.external-url` is set in Prometheus, Prometheus routes _all_ connections to that path. So if ingress points to `http://1.2.3.4/prometheus-unit-0/`, `web.external-url` is also set to `prometheus-unit-0/`. Ingress is not performing any path rewriting to push it back to `/`. 

In this case, even "in-namespace" requests to  `http://{prometheus_address}:9090` from inside the model/namespace will fail unless the path set in `web.external-url` is appended.

Failing to do so will cause lookups to `v1/api/query/....` to fail, and Grafana doesn't "see" Prometheus data.

If Loki operates the same way with `server.path-prefix= || http_path_prefix` (and it looks like it does), then the full URL/URI ("with" `http://` preferred, but not strictly required) should also be passed to `GrafanaSourceProvider` as the argument to `source_uri` when `web_external_url` has a value.

Reply:
@rbarry82 are we planning to rename [`source_url`](https://github.com/canonical/grafana-k8s-operator/blob/main/lib/charms/grafana_k8s/v0/grafana_source.py#L323) with `source_uri`?

Reply:
@Abuelodelanada we did ;-) https://github.com/canonical/grafana-k8s-operator/blob/ffa7b7f750a328bdf75cbbd3da300083bff5bf7f/lib/charms/grafana_k8s/v0/grafana_source.py#L30

Reply:
Way to sneak that one in 🤐 I was against that change, since naming it `source_uri` in the first place was a "hey, please include the schema when you pass this!" hint in the variable name. 

Sure the code checked if it included one, and added `http://` if it did not, but `_uri` was not an accident.

Reply:
URLs must have the scheme too. Every URL is a URI. The difference is that URIs does not have to be resolvable, URLs must.

Reply:
If the worry is about protocol-relative URLs, those are by definitions valid URIs too. It would make sense to me to validate the URL and ask it to be absolute or raise an exception.